rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+  machines/platforms, targeting yet a third machine/platform. This
+  does not necessarily imply different platforms; the machines could
+  be identical, except that the construction machine might faster,
+  have more RAM, etc.
+- **Mono-builds**: for lack of a better term. Same platform for all three
+  host roles.  A "native" build is a local mono-build.
+- **Cross-platform builds**: a/k/a "cross builds". The Target host
+  platform is different than the Bazel and Build host platforms.  A
+  cross-platform build could be either local or distributed.
+- **Canadian builds**: all three hosts are distinct platforms.  E.g. you
+  run Bazel on a mac, orchstrating build tools that run on a remote
+  linux machine, targeting Windows.
+
+### "Just works"
+
+The sense of "just works" is clear enough intuitively, but exactly
+what things/tasks should "just work" is a policy matter. For example,
+native builds clearly must just work, but which platforms should be
+supported is a policy decision. Should Windows native builds just
+work? If so, should the Cygwin toolchain be supported?  And so forth.
+
+What counts as just working in general:
+
+- Default local mono-builds ("native" builds) succeed on all supported
+  platforms; i.e. running build without passing any options by command
+  line or config file. Default settings just work, but may not be
+  optimal in every case.
+- Build refinements are easy and just work.  It must be easy for the
+  developer to express desired properties, and the ensuing build
+  configuration must succeed.  For example, optimization or debugging
+  flags.
+- Supported cross-platform builds must be easily expressible.  For example, it
+  should be possible to target Android by something like `$ bazel
+  build --platform=android_armv7`.
+- Configuration settings must be easily expressible either by command
+  line or via a config file (`.bazelrc`, `--bazelrc=file`).
+
+#### Integrated testing
+
+Bazel has builtin support for test targets.
+
+#### Remote caching and execution
+
+Bazel supports both of these, but they are outside the scope of this
+proposal.  Adding Bazel support would enable them.
+
+#### Continuous Integration support
+
+I have no experience with CI systems or processes, but Bazel claims to
+  have good support; e.g. see [Using Bazel in a continuous integration
+  system](https://blog.bazel.build/2016/01/27/continuous-integration.html)
+  and [Remote
+  cache](https://github.com/angular/angular/blob/master/docs/BAZEL.md#remote-cache).
+
+#### Maintainability and Extensibility
+
+It must be easy to maintain the build sytem.
+
+It should be easy to extend the system - for example, to add another
+target platform, toolchain, etc.  In particular, it should be easy to
+introduce new build tools.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+Platform support:
+  * Linux - required, I presume
+  * MacOS - my development environment
+  * Windows - this is a policy decision.  I can design for
```

Comment:
Windows builds should be out-of-scope for now. This can be a separate RFC and a project for later.

MacOS and Linux are both required :+1: